builder/codebuild/codebuild.go

```diff
@@ -0,0 +1,350 @@
+			return nil, errors.New("Log stream name could not be fetched, retry limit hit")
+		}
+    	
+    	if buildInfoResp.Builds[0].Logs == nil {
+
+    		time.Sleep(time.Second * 1)
+    		buildInfoResp, err = b.codebuild.BatchGetBuilds(params)
+
+    		if err != nil {
+    			return nil, err
+    		}
+
+
+    	} else {
+    		break;
+    	}
+
+    }
+
+    return &LogInfo{
+    	GroupName: *buildInfoResp.Builds[0].Logs.GroupName,
+    	StreamName: *buildInfoResp.Builds[0].Logs.StreamName,
+    }, nil
+
+}
+
+func (b *Builder) createProject(opts builder.BuildOptions, projectName string) (resp *codebuild.CreateProjectOutput, err error) {
+
+	log.Printf("Creating a new codebuild project: %s", projectName)
+
+	githubSource := fmt.Sprintf("https://github.com/%s", opts.Repository)
+	
+	buildParams := &codebuild.CreateProjectInput{
+	    Artifacts: &codebuild.ProjectArtifacts{
+	        Type:          aws.String("NO_ARTIFACTS"),
+	    },
+	    Environment: &codebuild.ProjectEnvironment{
+	        ComputeType: aws.String(b.ComputeType),
+	        Image:       aws.String(b.Image),
+	        Type:        aws.String("LINUX_CONTAINER"),
+	    },
+	    Name: aws.String(projectName),
+	    Source: &codebuild.ProjectSource{ 
+	        Type: aws.String("GITHUB"), 
+	        Auth: &codebuild.SourceAuth{
+	            Type:     aws.String("OAUTH"),
+	        },
+	        Location:  aws.String(githubSource),
+	    },
+	    ServiceRole:   aws.String(b.ServiceRole),
+	}
+
+	resp, err = b.codebuild.CreateProject(buildParams)
+
+	if err != nil {
+
+		return nil, err
+	}
+
+	return resp, nil
+}
+
+func (b *Builder) startBuild(opts builder.BuildOptions, projectName string) (resp *codebuild.StartBuildOutput, err error) {
+	
+	buildspec, err := b.generateBuildspec(opts)
+
+	if err != nil {
+		return
+	}
+
+	params := &codebuild.StartBuildInput{
+		ProjectName:   aws.String(projectName),
+		SourceVersion: aws.String(opts.Sha),
+		BuildspecOverride: aws.String(buildspec),
+	}
+
+	resp, err = b.codebuild.StartBuild(params)
+	return
+
+}
+
+
+func (b *Builder) generateBuildspec(opts builder.BuildOptions) (buildspec string, err error) {
+	
+	if b.Image != DefaultCodebuildImage {
+		err = errors.New("Please include a custom buildspec when using a different build Image")
+		return
+	}
+
+	params := BuildSpecInput{b, opts.Repository, opts.Sha, opts.Branch}
+	
+	specTemplate := `version: 0.1
+
+environment_variables:
+  plaintext:
+    DOCKER_USERNAME: {{.DockerUsername}}
+    DOCKER_PASSWORD: {{.DockerPassword}}
+
+phases:
+  pre_build:
+    commands:
+      - docker login -u ${DOCKER_USERNAME} -p ${DOCKER_PASSWORD}
+      - echo "Logged into Docker"
+      - docker pull "{{.Repository}}:${{.Branch}}" || docker pull "{{.Repository}}:master" || true
+      - echo "Pulled Image"
+  build:
+    commands:
+      - docker build -t "{{.Repository}}" .
+      - echo "Built Image with tag {{.Repository}}"
+      - docker tag "{{.Repository}}" "{{.Repository}}:{{.Branch}}"
```

Comment:
This should only tag if `BRANCH` is present: https://github.com/remind101/conveyor-builder/blob/master/bin/build#L96